### ISSUE TITLE
Implement hotspot feature

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-apple#170]
To allow the iPhone to be transformed in Wifi hotspot providing the same service as a standalone kiwix-serve. Now that the HTTP/kiwix-serve primitives are part of the libkiwix, this should be easy.

**Comment by kelson42** [comment]
Now that we have the same codebase for macOS and iOS, it should also work for macOS.

**Comment by kelson42** [comment]
@BPerlakiH Would that be doable? You can look on Android of Windows/linux version how it works.

**Comment by kelson42** [comment]
@BPerlakiH Any feedback here?

**Comment by BPerlakiH** [comment]
@kelson42 It seems it is doable. I did find the [Server class](https://libkiwix.readthedocs.io/en/stable/api/classkiwix_1_1Server.html#class-server) included in the libkiwix we are using. I can also see apps in the AppStore that are doing similar functionality (sharing local folders via a built in server).

**Comment by kelson42** [comment]
@BPerlakiH Great, feel free to start to implement it when you ar ready. But please (1) check how it work in kiwix android and kiwix desktop (2) propose a mockup or description about how it will work before real implementation.

**Comment by BPerlakiH** [comment]
I've already tried out the Android one.

**Comment by BPerlakiH** [comment]
@kelson42 This is my suggestion where we could put the Hotspot menu item within the Kiwix app:

-------
## iPad

![Image](https://github.com/user-attachments/assets/75bf0173-646a-4004-8d87-d92ab4da9062)

-------
## macOS

<img width="640" alt="Image" src="https://github.com/user-attachments/assets/842f74da-5746-4130-bffd-ee2474ee7c67" />

-------
## iPhone

![Image](https://github.com/user-attachments/assets/20948045-f8f4-4bb3-9a2c-c54469e16563)

-------

## The icon

I've used the standard wifi icon for this, but we can use any other [SF Symbols](https://developer.apple.com/design/human-interface-guidelines/sf-symbols) icon.

Some alternatives to consider:

<img width="320" alt="Image" src="https://github.com/user-attachments/assets/317fc68f-22f4-44f5-ae99-b149c5c99044" />

--------
<img width="320" alt="Image" src="https://github.com/user-attachments/assets/42372a84-4fb2-4fc4-ad0c-dd903c9c1370" />

-------
Note: on the system level there's a Personal Wifi hot-spot feature, which allows to share your Phone network over the wifi, that is using the hot-spot icon, as you can see on the control center:

![Image](https://github.com/user-attachments/assets/dad8af81-20c9-4192-acac-c2c626cb8085)

-----
## The name

I would keep it short "Hotspot", we need to fit it into the iPhone's tab bar, which will cut off / ellipse longer names.

## The content

As for the content itself under that menu item, the closest we currently have is the "Opened" menu, which lists the currently downloaded / added ZIM files. 
I need to come up with a transparent way to allow multi selection for them on all platforms.
Add add a start / stop button.

**Comment by BPerlakiH** [comment]
We could also consider to toggle the appearance of the icon similarly to system's control center, when the hotspot is turned on/off. Theoretically it should be possible.

**Comment by BPerlakiH** [comment]
Technical Question:
This is something I was curious about, so I tried it on Android.
I had 2 ZIM files that I started the hotspot with. While the hotspot was running, I deleted one of them. Now the hotspot main page was still showing 2 files. When tried to access the one I deleted, got the error back:

```
Internal Server Error

An internal server error occured. We are sorry about that :/

Error opening ZIM file: /storage/emulated/0/Kiwix/avp.stackexchange.com_en_all_2024-10.zim
```

Is that a good enough way to go about this edge case?

**Comment by kelson42** [comment]
> Technical Question: This is something I was curious about, so I tried it on Android. I had 2 ZIM files that I started the hotspot with. While the hotspot was running, I deleted one of them. Now the hotspot main page was still showing 2 files. When tried to access the one I deleted, got the error back:
> 
> ```
> Internal Server Error
> 
> An internal server error occured. We are sorry about that :/
> 
> Error opening ZIM file: /storage/emulated/0/Kiwix/avp.stackexchange.com_en_all_2024-10.zim
> ```
> 
> Is that a good enough way to go about this edge case?

No, please open issue at kiwix-android. It should dynamically removed the file from the hotspot (and restart it if necessary). Actually looks like a bug in libkiwix.

**Comment by kelson42** [comment]
@BPerlakiH Thx, but how would concretly looks like the user journey to activate/deactivate the hotspot. I see the kiwix sidebar menu entry which is OK, but for the rest it has to be defined.

**Comment by BPerlakiH** [comment]
Issue reported on Android: https://github.com/kiwix/kiwix-android/issues/4341

**Comment by BPerlakiH** [comment]
@kelson42 this is how I solved the UI part:

The overall concept is that we can select multiple ZIM files, once the server is running we cannot change the selection any more.

## macOS

https://github.com/user-attachments/assets/a65b3bf4-f1a7-40d1-875f-45c1905089f8


----
## iPhone

https://github.com/user-attachments/assets/a81d2a77-7fa3-48f6-86af-43a93b208f41

---- 
## iPad

https://github.com/user-attachments/assets/c72c633f-b045-48ec-939a-ae29c040f215

**Comment by BPerlakiH** [comment]
BTW: I am also thinking about how to add the QR code and the IP address once the server is started.

**Comment by BPerlakiH** [comment]
@kelson42 I managed to get the Hotspot up and running. It does seem to work really well so far!
I need to add the things mentioned above, and the port settings.

One thing I have noticed though is, that while the ZIM files itself do support dark mode, the hotspot server itself does not:

<img width="1016" height="581" alt="Image" src="https://github.com/user-attachments/assets/61e10185-450a-4e98-92b5-59d778d32cca" />
<img width="1088" height="416" alt="Image" src="https://github.com/user-attachments/assets/8644c7ba-c472-44c8-a307-f72defe24829" />

**Comment by kelson42** [comment]
@BPerlakiH Great news, I can not wait to have a demonstration with you or a screencast. For the dark mode, this is normal, it comes then from the OS/browser them settings.

**Comment by BPerlakiH** [comment]
@kelson42 What I meant is that the CSS file, which is part of the server itself, is not respecting system dark mode. If I am not mistaken this is the one:
https://github.com/kiwix/libkiwix/blob/699cbccf3864946b3b323243350483724cab13f0/static/skin/kiwix.css

**Comment by BPerlakiH** [comment]
macOS with QR code:

<img width="1281" height="830" alt="Image" src="https://github.com/user-attachments/assets/f6245d78-704c-4925-b4a4-da79029c572e" />

**Comment by BPerlakiH** [comment]
and for iOS:

<img width="295" height="640" alt="Image" src="https://github.com/user-attachments/assets/c2f4df1a-d71e-4b30-ac25-c5eee03086fe" />

**Comment by BPerlakiH** [comment]
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/ff32d92e-6549-42e1-ad57-a5a7437f7beb" />

**Comment by BPerlakiH** [comment]
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/ac33401c-be32-4d5c-9b05-67e28dcab0f3" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/0721558b-5af4-45ca-8c64-ce47346b3169" />
<img width="295" height="640" alt="Image" src="https://github.com/user-attachments/assets/b21c7311-3894-4419-a326-e4c9d91a9cf8" />
<img width="295" height="640" alt="Image" src="https://github.com/user-attachments/assets/6cb92231-5dfb-44dd-a18e-96af056ca2d9" />
<img width="295" height="640" alt="Image" src="https://github.com/user-attachments/assets/e349f552-581a-4723-b860-824c30c51669" />

**Comment by kelson42** [comment]
I reopen the issue to polish the feature, but I'm still working on this.

First points to fix are:
* We should better to propose an easy way to the user to share the URL and the QR code
* We should find a way to create a WiFI access point (to oppose to join the same network used by the device)

**Comment by BPerlakiH** [comment]
@kelson42 these 2 PRs are addressing the above 2 points:
[Update Hotspot instructions #1238](https://github.com/kiwix/kiwix-apple/pull/1238)
[170 hotspot share #1240](https://github.com/kiwix/kiwix-apple/pull/1240)

**Comment by kelson42** [comment]
> I reopen the issue to polish the feature, but I'm still working on this.
> 
> First points to fix are:
> 
> * We should better to propose an easy way to the user to share the URL and the QR code
> * We should find a way to create a WiFI access point (to oppose to join the same network used by the device)

The second point has not been treated AFAIK

**Comment by kelson42** [comment]
> > I reopen the issue to polish the feature, but I'm still working on this.
> > First points to fix are:
> > 
> > * We should better to propose an easy way to the user to share the URL and the QR code
> > * We should find a way to create a WiFI access point (to oppose to join the same network used by the device)
> 
> The second point has not been treated AFAIK, I see no way in Kiwix to start the access point.

**Comment by BPerlakiH** [comment]
@kelson42 At this point it seems to me that the real networking part of this (how to connect 2 or more devices) is on the system level.
For iOS it is documented by Apple:
https://support.apple.com/en-us/111785

For macOS it is either by [Internet Sharing](https://support.apple.com/en-il/guide/mac-help/mchlp1540/mac), or if there's no internet at all, it's a bit more tricky, but can be done this way (I tested and it is working using the "How to Create a Secure Ad Hoc Network" via terminal part):
https://www.makeuseof.com/how-to-create-a-secure-ad-hoc-network-in-macos/

**Comment by kelson42** [comment]
@BPerlakiH Please make a mockup of the wokflow to activate the access point for both iOS and macOS

**Comment by BPerlakiH** [comment]
@kelson42 I suppose this is the one you had in mind from Android:

![Image](https://github.com/user-attachments/assets/1970a171-f4b6-4066-9c20-eb98e18fd911)

The issue on the iOS/mac side is that Apple only allows to redirect to the [Kiwix section of the Settings](https://developer.apple.com/documentation/uikit/uiapplication/opensettingsurlstring), not to the main Settings Page, not the Network Settings. None of that is allowed by "App Store rules".

**Comment by BPerlakiH** [comment]
In our case we could only redirect the user here (on iOS):

![Image](https://github.com/user-attachments/assets/43cb355c-cf74-4ced-9fed-eba7f7478e9d)

**Comment by kelson42** [comment]
> In our case we could only redirect the user here (on iOS):
> 
> ![Image](https://github.com/user-attachments/assets/43cb355c-cf74-4ced-9fed-eba7f7478e9d)

and...!?! Does that mean this is not possible? And macOS? Please think the problem trough and propose a solution.

**Comment by BPerlakiH** [comment]
Yes, it means it is not possible to redirect the user to those system settings directly.

In this regard macOS is even worse, because "https://developer.apple.com/documentation/uikit/uiapplication/opensettingsurlstring" is not available for macOS, as there are no app specific settings on the system level.

=====

I have been thinking about it, and the most straightforward we can do is to inform the user what options do work, which we do in this message:

<img width="259" height="167" alt="Image" src="https://github.com/user-attachments/assets/78f43697-8a2d-4ecb-bcb1-8379879a7167" />

Where "Personal Hotspot" is the official name of the Settings option, the user should reach for:

![Image](https://github.com/user-attachments/assets/1c12f012-3fbd-470f-96e3-870c6a0b50f2)

**Comment by kelson42** [comment]
I have no "personal hotapot" on my iPad!? What about the feedback: do we have a programmatic way to known that the device is in access point mode?

**Comment by BPerlakiH** [comment]
Hmm that's strange, you should have one: https://support.apple.com/en-us/111785

**Comment by kelson42** [comment]
See by yourself
![image](https://github.com/user-attachments/assets/7a47fb7c-dc5b-47d8-8328-bebd95fd3d7f)

What about my next question?

**Comment by rgaudin** [comment]
Personnal Hotspot on iOS/iPadOS refers to sharing a mobile connection over WiFi. If the device is WiFi only (no SIM/eSIM), then it's not possible to share it obviously.

**Comment by kelson42** [comment]
> Personnal Hotspot on iOS/iPadOS refers to sharing a mobile connection over WiFi. If the device is WiFi only (no SIM/eSIM), then it's not possible to share it obviously.

"personal hotspot" is a tethering feature and we should probably consider it as having nothing to do with Kiwix then. Is that correct? Or should we consider tethering as a way to access the kiwix hotspot (if it work at all!)?

**Comment by rgaudin** [comment]
Nothing to do with Kiwix indeed. Only (another) name collision

**Comment by kelson42** [comment]
Should I conclude that there is no way to propose an access point for Kiwix, like on Kiwix Android?

**Comment by rgaudin** [comment]
I don't know what the feature does an Android and can't check at the moment but AFAIK, there is no system-provided SDK on macOS nor iOS to automatically create a WiFi network with an associated DHCP server (a _hotspot_).
macOS also has an internet sharing feature but it's also for one iface to another one, is not flexible and has no API.

**Comment by BPerlakiH** [comment]
In terms of what is possible for the user, I have tested the tethering (Personal Hotspot) way on iOS, and internet sharing on macOS, both can be used for Kiwix purposes. I did the summary of those up here: https://github.com/kiwix/kiwix-apple/issues/170#issuecomment-3085462857

There's no (Apple approved) programatic way to create the Hotspot from the Kiwix app, (even the not approved / deprecated one was not working for me).

The closest I've found in terms of detection is the current connection type, but tethering / "Personal Hotspot" is not included on that either:
https://developer.apple.com/documentation/network/nwinterface/interfacetype#Interface-Types

**Comment by rgaudin** [comment]
@kelson42 should we close this or at least remove it from 3.10 milestone?
